tower-balance/src/p2c/service.rs

```diff
@@ -21,26 +23,47 @@ use tower_service::Service;
 /// [finagle]: https://twitter.github.io/finagle/guide/Clients.html#power-of-two-choices-p2c-least-loaded
 /// [p2c]: http://www.eecs.harvard.edu/~michaelm/postscripts/handbook2001.pdf
 #[derive(Debug)]
-pub struct Balance<D: Discover> {
+pub struct Balance<D: Discover, Req> {
     discover: D,
 
-    endpoints: IndexMap<D::Key, D::Service>,
+    ready: IndexMap<D::Key, D::Service>,
+
+    unready: stream::FuturesUnordered<CancelReady<D::Key, D::Service, Req>>,
+    cancelations: IndexMap<D::Key, oneshot::Sender<()>>,
 
     /// Holds an index into `endpoints`, indicating the service that has been
     /// chosen to dispatch the next request.
-    ready_index: Option<usize>,
+    next_ready_index: Option<usize>,
 
     rng: SmallRng,
 }
 
-impl<D: Discover> Balance<D> {
+#[derive(Debug)]
+struct CancelReady<K, S, Req> {
+    key: Option<K>,
+    cancel: oneshot::Receiver<()>,
+    ready: tower_util::Ready<S, Req>,
+}
+
+enum Error<E> {
+    Inner(E),
+    Canceled,
+}
+
+impl<D, Req> Balance<D, Req>
+where
+    D: Discover,
+    D::Service: Service<Req>,
```

Comment:
I don't think you need this bound here?

Reply:
It's needed so that `UnreadyService<..>: Future`, which is required by `FuturesUnordered`